### DISCUSSION TITLE
X-Y Real-Time Plot without Time Axis

### DISCUSSION THREAD

**Discussion by RagnvaldIV** [collin80/SavvyCAN#595] [Ideas]
Hello!

I'd like to use SavvyCAN to do a real-time plot that is **not** time-base. I see the [Graphing Window documentation](https://www.savvycan.com/docs/graphwindow.html) that lets you plot in real-time but you only select the Y-axis and the X-axis is always time. I'd like to be able to select the X-axis as another signal in the same message (I can see some complication if they are not perfectly time-sync'd, maybe this is future work with matching the closest timed variables?).

A use-case example of this is mapping the engine/motor RPM to groundspeed to visually identify/indicate gear ratios.

Thanks!

**Comment by collin80** [comment]
Hmm... It's an interesting idea. I suppose what would have to happen is that you key the graph update to one of the axes, say the value set to the X axis, and then take the most recent Y axis value and plot the <X,Y> point as either a point or line segment. This could even be done from existing captures - just still grab every time the given trigger axis has a new value and use the most recent Y value then plot the point. No guarantees but this may be something I'll end up adding in the future.

**Reply by RagnvaldIV** [reply]
Awesome, thank you!

I agree that your proposed update method would be suitable to match latest Y with latest X. 

For now I managed to implement this in python with python-can, can-tools, and matplotlib

**Comment by mitchdetailed** [comment]
this is somewhat why i wrote a python script to take socketcan input and a
DBC to decode the information and export a CSV to live analyze in megalog
viewer HD. maths on channels for various purposes..

On Tue, Feb 28, 2023 at 7:55 PM Collin Kidder ***@***.***>
wrote:

> Hmm... It's an interesting idea. I suppose what would have to happen is
> that you key the graph update to one of the axes, say the value set to the
> X axis, and then take the most recent Y axis value and plot the <X,Y> point
> as either a point or line segment. This could even be done from existing
> captures - just still grab every time the given trigger axis has a new
> value and use the most recent Y value then plot the point. No guarantees
> but this may be something I'll end up adding in the future.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/collin80/SavvyCAN/discussions/595#discussioncomment-5161216>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AI62NLVBM24NSPZVHKUUCQ3WZ2T75ANCNFSM6AAAAAAVKCBYKQ>
> .
> You are receiving this because you are subscribed to this thread.Message
> ID: ***@***.***>
>


-- 
Minton Performance
615-977-9787

**Comment by MotorvateDIY** [comment]
I've done this, but used an ESP32 on the CAN bus.
It reads the engine RPM and vehicle speed from the CAN bus broadcasts, calculates the current gear, then sends that back out on the CAN bus. (using the gauge cluster CAN ID that displays 1-6 next to the odometer, as a gear position indicator)
